File: 06_Linked_List/07_Linked_List_Cycle/0141-linked-list-cycle.rs
/*
Problem: LeetCode 141 - Linked List Cycle

Key Idea:
The key idea is to use two pointers, one slow and one fast, to detect the presence of a cycle in the linked list.

Approach:
1. Initialize two pointers, `slow` and `fast`, to the head of the linked list.
2. Traverse the linked list with the pointers:
   - Move `slow` one step at a time.
   - Move `fast` two steps at a time.
3. If there is a cycle in the linked list, `fast` will eventually catch up to `slow`, and they will meet at some point in the cycle.
4. If there is no cycle, `fast` will reach the end of the list (i.e., become None) before `slow`.
5. Return true if there is a cycle, and false if there is no cycle.

Time Complexity:
O(N), where N is the number of nodes in the linked list. In the worst case, both pointers traverse the entire list once.

Space Complexity:
O(1), as we use a constant amount of extra space for the two pointers.
*/

// No option to solve this problem in Rust

File: 06_Linked_List/07_Linked_List_Cycle/0141-linked-list-cycle.py
"""
Problem: LeetCode 141 - Linked List Cycle

Key Idea:
To detect a cycle in a linked list, we can use the Floyd's Tortoise and Hare algorithm. We maintain two pointers, 'slow' and 'fast', where 'slow' moves one step at a time and 'fast' moves two steps at a time. If there is a cycle in the linked list, the two pointers will eventually meet at some point. If there is no cycle, the 'fast' pointer will reach the end of the list. 

Time Complexity:
The time complexity of this solution is O(n), where n is the number of nodes in the linked list. In the worst case, both pointers traverse the entire list.

Space Complexity:
The space complexity is O(1), as no extra space is used other than a few variables to keep track of nodes and pointers.
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        if not head or not head.next:
            return False

        slow = head
        fast = head.next

        while slow != fast:
            if not fast or not fast.next:
                return False
            slow = slow.next
            fast = fast.next.next

        return True
